src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBMetricsProvider.cs

```
@@ -0,0 +1,135 @@
+            { "Request rate is large", string.Empty },
+            { "PartitionKey value must be supplied for this operation.", "We do not support lease containers with partitions at this time. Please create a new lease collection without partitions." },
+            { "The remote name could not be resolved:", string.Empty },
+            { "Owner resource does not exist", string.Empty },
+            { "The specified document collection is invalid", string.Empty }
+        };
+
+        public CosmosDBMetricsProvider(ILogger logger, Container monitoredContainer, Container leaseContainer, string processorName)
+        {
+            _logger = logger;
+            _monitoredContainer = monitoredContainer;
+            _leaseContainer = leaseContainer;
+            _processorName = processorName;
+        }
+
+        public async Task<CosmosDBTriggerMetrics> GetMetricsAsync()
+        {
+            int partitionCount = 0;
+            long remainingWork = 0;
+
+            try
+            {
+                List<ChangeFeedProcessorState> partitionWorkList = new List<ChangeFeedProcessorState>();
+                ChangeFeedEstimator estimator = _monitoredContainer.GetChangeFeedEstimator(_processorName, _leaseContainer);
+                using (FeedIterator<ChangeFeedProcessorState> iterator = estimator.GetCurrentStateIterator())
+                {
+                    while (iterator.HasMoreResults)
+                    {
+                        FeedResponse<ChangeFeedProcessorState> response = await iterator.ReadNextAsync();
+                        partitionWorkList.AddRange(response);
+                    }
+                }
+
+                partitionCount = partitionWorkList.Count;
+                remainingWork = partitionWorkList.Sum(item => item.EstimatedLag);
+            }
+            catch (Exception e) when (e is CosmosException || e is InvalidOperationException)
```

We want to handle all `CosmosExceptions` but throw `InvalidOperationException` ?

```suggestion
            catch (Exception e) when (e is CosmosException)
```

---

I believe InvalidOperationException is meant to be handled in TryHandleCosmosException -- there are specific unit tests in CosmosDBMetricsProviderTests.

---

The SDK throws CosmosException. If the consumer of this API understands InvalidOperationException that would be a contract between this code and that consumer, I cannot chip in here because we are neither the consumer nor the authors or this code originally.